src/plugins/embeddable/public/lib/embeddables/embeddable_renderer.tsx

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useState } from 'react';
+import { EmbeddableInput, IEmbeddable } from './i_embeddable';
+import { EmbeddableRoot } from './embeddable_root';
+import { EmbeddableFactory } from './embeddable_factory';
+import { ErrorEmbeddable } from './error_embeddable';
+
+/**
+ * This type is needed for strict, one of, public api
+ * input is required only when rendering factory
+ */
+export type EmeddableRendererProps<I extends EmbeddableInput> =
+  | { input: I; onInputUpdated?: (newInput: I) => void; factory: EmbeddableFactory<I> }
+  | { input?: I; onInputUpdated?: (newInput: I) => void; embeddable: IEmbeddable<I> };
+
+/**
+ * This one is for internal implementation
+ */
+interface InnerProps {
+  input?: EmbeddableInput;
+  onInputUpdated?: (newInput: EmbeddableInput) => void;
+  factory?: EmbeddableFactory;
+  embeddable?: IEmbeddable;
+}
+
+export const EmbeddableRenderer = <I extends EmbeddableInput>(
```

Comment:
I think we need to be extra clear that this renderer shouldn't be used to render embeddable children inside containers because it will blow away implicit state.  Maybe add some tsdocs here, mention in the README and the examples UI?  In the readme you could probably get into a little more detail.

## Embeddable containers and inherited input state 

 `updateInput` is typed as `updateInput(input: Partial<EmbeddableInput>)`.  Notice it's _partial_.  This is to support the use case of inherited state when an embeddable is inside a container.

If you are simply rendering an embeddable, it's no problem to do something like:
```
// Notice this isn't a partial so it'll be the entire state.
const input: EmbeddableInput = this.state.input
embeddable.updateInput(input);
```
However when you are dealing with _containers_, you want to be sure to **only pass into `updateInput` the actual state that changed**.  This is because calling `child.updateInput({ foo })` will make `foo` _explicit_ state. It cannot be inherited from it's parent.

For example, on a dashboard, the time range is _inherited_ by all children, _unless_ they had their time range set explicitly.  This is how "per panel time range" works.  That action calls `embeddable.updateInput({ timeRange })`, and the time range will no longer be inherited from the container.  

### Why is this important?

A common mistake is always passing in the full state.  If you do this, all of a sudden you will lose the inheritance of the container state.

**Don't do**
```ts
// Doing this will make it so this embeddable inherits _nothing_ from its container. No more time range updates
// when the user updates the dashboard time range!
embeddable.updateInput({ ...embeddable.getInput(), foo: 'bar' });
```

**Do**
```ts
embeddable.updateInput({ foo: 'bar' });
```

lmk if that makes sense!

Reply:
This is great! this explanation even helpful to me.. 

I put a shorter version of it into TSDoc and linked to this comment. 
But not sure where to put the whole version. 

it seems like we are missing a place where to put such kind of guides or detailed explanations? 

Not sure I want to add it into `src/plugins/embeddable/README.md` because how minimalistic it it now. 

I am thinking of 

`src/plugins/embeddable/docs/embeddable_containers_and_inherited_state.md` and then Link from `src/plugins/embeddable/README.md` -> `src/plugins/embeddable/docs/embeddable_containers_and_inherited_state.md`

?

Reply:
sgtm